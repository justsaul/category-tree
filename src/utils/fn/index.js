export const range = (len) => [...Array(len).keys()]
