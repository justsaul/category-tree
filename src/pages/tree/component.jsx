import React from 'react'
import Provider from '../../components/tree/provider'

import { TYPE as TREE_TYPE } from '../../utils/tree'

const TreePage = (props) => (
    <>
        <Provider {...props} type={TREE_TYPE.RECURSIVE}/>
        <Provider {...props} type={TREE_TYPE.ITERATIVE}/>
    </>
)

export default TreePage
