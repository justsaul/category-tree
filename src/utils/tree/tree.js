import { range } from '../fn'

export const TYPE = {
  RECURSIVE: 'recursive',
  ITERATIVE: 'iterative',
}

export class Tree {
  constructor(value) {
    this.value = value
    this.children = []
  }

  addChildren(value) {
    value instanceof Tree
      ? this.children.push(value)
      : this.children.push(new Tree(value))

    return this
  }

  getValue() {
    return this.value
  }

  getChildren() {
    return this.children
  }

  insert(insertable, atNode, tree=this) {
    if (tree.value === atNode.value) {
      tree.addChildren(insertable)
      return true
    }

    if (~tree.children.length) {
      range(tree.children.length).forEach((i) => {
        this.insert(insertable, atNode, tree.children[i])
      })
    }
    return false
  }
}
