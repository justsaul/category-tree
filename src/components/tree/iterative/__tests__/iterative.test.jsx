import React from 'react'
import { shallow } from 'enzyme'

import { Tree } from '../../../../utils/tree'
import { TreeIterative } from '../'

describe('<TreeIterative/>', () => {
    it('should render without exploding', () => {
        const Lambda = new Tree('Lambda')
        const wrapper = shallow(
            <TreeIterative tree={Lambda} update={jest.fn()}/>
        )

        expect(wrapper).toMatchSnapshot()
    })
    it('should display elements nested', () => {
        const Lambda = new Tree('Lambda')
        Lambda.addChildren(new Tree('Children'))
        
        const wrapper = shallow(
            <TreeIterative tree={Lambda} update={jest.fn()} />
        )

        expect(wrapper.children().length).toEqual(2)
    })
})
