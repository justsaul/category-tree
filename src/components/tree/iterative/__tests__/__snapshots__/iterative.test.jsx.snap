// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TreeIterative/> should render without exploding 1`] = `ShallowWrapper {}`;
